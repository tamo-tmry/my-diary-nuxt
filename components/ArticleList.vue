<template>
  <div>
    <v-list v-if="hasArticle">
      <v-list-item v-for="article in articles" :key="article.id" :to="`/article/${article.id}`">
        <v-list-item-title>{{ article.title }}</v-list-item-title>
        <v-list-item-subtitle>{{ article.content }}</v-list-item-subtitle>
      </v-list-item>
    </v-list>
    <p v-else>投稿がありません。</p>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'

export default Vue.extend({
  props: {
    articles: {
      type: Array as PropType<any[]>, // TODO: 型定義修正
      required: true,
    },
  },
  computed: {
    hasArticle() {
      return Boolean(this.articles.length)
    },
  },
})
</script>
