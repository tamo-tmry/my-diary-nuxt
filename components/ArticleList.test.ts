import { render, screen } from '@testing-library/vue'
import Vue from 'vue'
import Vuetify from 'vuetify'
import VueRouter from 'vue-router'
import ArticleList from './ArticleList.vue'

Vue.use(Vuetify)
Vue.use(VueRouter) // VueにVueRouterをインストールします

const router = new VueRouter() // 新しいRouterインスタンスを作成します

test('一覧表示される', () => {
  const vuetify = new Vuetify()

  const articles = [
    {
      id: 'article-id-1',
      title: 'article-title-1',
      content: 'article-content-1',
    },
    {
      id: 'article-id-2',
      title: 'article-title-2',
      content: 'article-content-2',
    },
    {
      id: 'article-id-3',
      title: 'article-title-3',
      content: 'article-content-3',
    },
  ]

  render(ArticleList, {
    props: {
      articles,
    },
    vuetify,
    routes: router,
  })
  const list = screen.getByRole('list')
  expect(list).toBeInTheDocument()
})

test('articlesの数だけ一覧表示される', () => {
  const vuetify = new Vuetify()

  const articles = [
    {
      id: 'article-id-1',
      title: 'article-title-1',
      content: 'article-content-1',
    },
    {
      id: 'article-id-2',
      title: 'article-title-2',
      content: 'article-content-2',
    },
    {
      id: 'article-id-3',
      title: 'article-title-3',
      content: 'article-content-3',
    },
  ]

  render(ArticleList, {
    props: {
      articles,
    },
    vuetify,
    routes: router,
  })
  expect(screen.getAllByRole('listitem')).toHaveLength(articles.length)
})

test('投稿がない場合、「投稿がありません。」が表示される', () => {
  const vuetify = new Vuetify()

  render(ArticleList, {
    props: {
      articles: [],
    },
    vuetify,
    routes: router,
  })
  const list = screen.queryByRole('list')
  expect(list).not.toBeInTheDocument()
  expect(list).toBeNull()
  expect(screen.getByText('投稿がありません。')).toBeInTheDocument()
})

test('IDに基づいたリンクが表示される', () => {
  const vuetify = new Vuetify()

  const articles = [
    {
      id: 'article-id-1',
      title: 'article-title-1',
      content: 'article-content-1',
    },
  ]

  render(ArticleList, {
    props: {
      articles,
    },
    vuetify,
    routes: router,
  })
  expect(screen.getByRole('listitem')).toHaveAttribute(
    'href',
    '#/article/article-id-1'
  )
})
