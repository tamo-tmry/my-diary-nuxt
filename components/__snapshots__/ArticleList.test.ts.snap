// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: 一覧表示される 1`] = `
<div>
  <div>
    <div
      class="v-list v-sheet theme--light"
      role="list"
    >
      <a
        class="v-list-item v-list-item--link theme--light"
        href="#/article/article-id-1"
        role="listitem"
        tabindex="0"
      >
        <div
          class="v-list-item__title"
        >
          article-title-1
        </div>
         
        <div
          class="v-list-item__subtitle"
        >
          article-content-1
        </div>
      </a>
      <a
        class="v-list-item v-list-item--link theme--light"
        href="#/article/article-id-2"
        role="listitem"
        tabindex="0"
      >
        <div
          class="v-list-item__title"
        >
          article-title-2
        </div>
         
        <div
          class="v-list-item__subtitle"
        >
          article-content-2
        </div>
      </a>
      <a
        class="v-list-item v-list-item--link theme--light"
        href="#/article/article-id-3"
        role="listitem"
        tabindex="0"
      >
        <div
          class="v-list-item__title"
        >
          article-title-3
        </div>
         
        <div
          class="v-list-item__subtitle"
        >
          article-content-3
        </div>
      </a>
    </div>
  </div>
</div>
`;
